pkg/syncer/volume_health.go

```diff
@@ -77,58 +77,75 @@ func csiGetVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface
 		}
 	}
 
-	for _, vol := range queryResult.Volumes {
-		log.Debugf("Volume %q Health Status %q", vol.VolumeId.Id, vol.HealthStatus)
+	for volID, pvc := range volumeHandleToPvcMap {
+		foundVol := false
+		for _, vol := range queryResult.Volumes {
+			log.Debugf("Volume %q Health Status %q", vol.VolumeId.Id, vol.HealthStatus)
+			if volID == vol.VolumeId.Id {
+				log.Debugf("csiGetVolumeHealthStatus: Found pvc %q for volume %q", pvc, vol.VolumeId.Id)
+				foundVol = true
+				// only update PVC health annotation if the HealthStatus of volume is not "unknown"
+				if vol.HealthStatus != string(pbmtypes.PbmHealthStatusForEntityUnknown) {
+					volHealthStatus, err := common.ConvertVolumeHealthStatus(vol.HealthStatus)
+					if err != nil {
+						log.Errorf("csiGetVolumeHealthStatus: invalid health status %q for volume %q", vol.HealthStatus, vol.VolumeId.Id)
+					}
+					updateVolumeHealthStatus(ctx, k8sclient, pvc, volHealthStatus)
+				}
+			}
+		}
+		if !foundVol {
+			// Set volume health status as "Inaccessible" when PVC is not found in CNS.
+			// When a Datastore is removed from VC (like vSAN direct disk decommisson with noAction does), the CNS Volumes
+			// on that Datastore are eventually removed from CNS DB, but the PVCs still remain in the K8S cluster.
+			updateVolumeHealthStatus(ctx, k8sclient, pvc, common.VolHealthStatusInaccessible)
+		}
+	}
+	log.Infof("GetVolumeHealthStatus: end")
+}
 
-		if pvc, ok := volumeHandleToPvcMap[vol.VolumeId.Id]; ok {
-			log.Debugf("csiGetVolumeHealthStatus: Found pvc %q for volume %q", pvc, vol.VolumeId.Id)
+func updateVolumeHealthStatus(ctx context.Context, k8sclient clientset.Interface, pvc *v1.PersistentVolumeClaim, volHealthStatus string) {
+	log := logger.GetLogger(ctx)
+	log.Infof("updateVolumeHealthStatus: start")
```

Comment:
Since this is called for each PVC, I would recommend to remove it.

Reply:
Done